### PR TITLE
Append unit to prometheus metric names

### PR DESCRIPTION
**Pull request by psx95**

### CODE REVIEW

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+    // Converting abbreviated unit names to full names
+    return cleanUpString(
+        PROMETHEUS_UNIT_MAP.getOrDefault(convertedMetricUnitName, convertedMetricUnitName));
+  }
+
+  private static String convertRateExpressedToPrometheusUnit(String rateExpressedUnit) {
+    String[] rateEntities = rateExpressedUnit.split("/", 2);
+    // Only convert rate expressed units if it's a valid expression
+    if (rateEntities[1].equals("")) {
```
Comment (jsuereth) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java:139]: Document that you require a "/" to exist in the string before calling the function.

As an alternative, you could move the "/" check here as an early exit that just returns the passed in string.

Comment (psx95) [reply exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java:139]: Right, this sounds good - this way method will be self-sufficient even though its a private one.

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+ * @see <a href="https://prometheus.io/docs/practices/naming/#base-units">Prometheus best practices
+ *     for units</a>
+ */
+public final class PrometheusUnitsHelper {
+
+  private static final Pattern INVALID_CHARACTERS_PATTERN = Pattern.compile("[^a-zA-Z0-9]");
+  private static final String CHARACTERS_BETWEEN_BRACES_REGEX =
+      "\\{(.*?)}"; // matches all characters between {}
+
+  private static final Map<String, String> PROMETHEUS_UNIT_MAP =
```
[the remark this answers is not in the document]
Comment (psx95) [reply exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java:33]: Discussed offline, switch seems better for performance.

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    convertedMetricUnitName = removeUnitPortionInBrackets(convertedMetricUnitName);
+
+    // Handling for the "per" unit(s), e.g. foo/bar -> foo_per_bar
+    convertedMetricUnitName = convertRateExpressedToPrometheusUnit(convertedMetricUnitName);
+
+    // Converting abbreviated unit names to full names
+    return cleanUpString(getPrometheusUnit(convertedMetricUnitName));
+  }
+
+  private static String convertRateExpressedToPrometheusUnit(String rateExpressedUnit) {
```
Comment (jsuereth) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java:64]: While not strictly required, can you add javadoc to this method?

It helps a lot for others following the algorithm.

**Review by jsuereth — APPROVED** [review]
Down to minor commments.  

Overall LGTM.

High level - we should probably optimise the string manipulation to involve less allocations over time.  We can check w/ @jack-berg on whether or not we have a prometheus-exporter benchmark where we could measure allocations, etc.  That can be follow on work here.

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+ * A utility class that contains helper function(s) to perform conversion from OTLP to Prometheus
+ * units.
+ *
+ * @see <a
+ *     href="https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#units-and-base-units">OpenMetrics
+ *     specification for units</a>
+ * @see <a href="https://prometheus.io/docs/practices/naming/#base-units">Prometheus best practices
+ *     for units</a>
+ */
+public final class PrometheusUnitsHelper {
```
Comment (mateuszrzeszutek) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java:21]: This probably shouldn't be a part of the public API
```suggestion
final class PrometheusUnitsHelper {
```

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+  private static final Pattern INVALID_CHARACTERS_PATTERN = Pattern.compile("[^a-zA-Z0-9]");
+  private static final String CHARACTERS_BETWEEN_BRACES_REGEX =
```
Comment (mateuszrzeszutek) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java:24]: Please use `Pattern` for all regexes

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Serializer.java [hunk 6]
```diff
@@ -650,8 +651,15 @@ static Collection<? extends PointData> getPoints(MetricData metricData) {
     return Collections.emptyList();
   }
 
-  private static String metricName(String rawMetricName, PrometheusType type) {
-    String name = NameSanitizer.INSTANCE.apply(rawMetricName);
+  private static String metricName(MetricData rawMetric, PrometheusType type) {
+    String name = NameSanitizer.INSTANCE.apply(rawMetric.getName());
```
Comment (mateuszrzeszutek) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Serializer.java:655]: If you move this line below the next `if` you can probably get rid of the `cleanUpString` method in `PrometheusUnitsHelper` -- no need to perform sanitization twice after all. (And the performance will benefit from the cache in `NameSanitizer` as well)

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+   * Prometheus compliant. This method also removes leading and trailing underscores.
+   *
+   * @param string The string input that needs to be made Prometheus compliant.
+   * @return the cleaned-up Prometheus compliant string.
+   */
+  private static String cleanUpString(String string) {
+    String prometheusCompliant =
+        INVALID_CHARACTERS_PATTERN.matcher(string).replaceAll("_").replaceAll("[_]{2,}", "_");
+    prometheusCompliant = prometheusCompliant.replaceAll("_+$", ""); // remove trailing underscore
+    prometheusCompliant = prometheusCompliant.replaceAll("^_+", ""); // remove leading underscore
```
Comment (mateuszrzeszutek) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java:107]: What's the reason for these replacements? Prometheus does not seem to forbid double or trailing underscores in its spec; and it does not seem that the previous operations on the unit are likely to produce trailing or duplicated underscores.

Comment (psx95) [reply exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java:107]: The spec linked in the PR  description states - 
```
 Multiple consecutive `_` characters MUST be replaced with a single `_` character
```

Based on that I removed leading `_`, because the serializer was already adding it.  
```
name = name + "_" + prometheusEquivalentUnit;
```

But looking at this again, I think this would be confusing to do in the `PrometheusUnitHelper`.  I will remove this, also the spec does not explicitly mention that the final unit name cannot have a trailing `_`, so I think leading `_` is fine. I will make changes. 

Also, the current name sanitizer does not seem to remove consecutive `_`, so based on spec this needs to be added as well.

#### TL;DR
 - The PrometheusUnitHelper will just return the computed unit (with illegal characters replaced with `_`) without any additional cleaning up. 
 - The Sanitizing will be done by the serializer right before returning. The sanitizing will only take care of replacing any remaining Illegal characters with `_` and then replacing consecutive `_` with a single underscore.

Comment (psx95) [reply exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java:107]: I would like to mention here that while not explicitly stated in the spec, this is how opentelemetry-collector [works](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9a9d4778bbbf242dba233db28e2fbcfda3416959/pkg/translator/prometheus/normalize_name.go#L277), should we try to keep this in sync with how the collector works ?

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+   * @param string The string input that needs to be made Prometheus compliant.
+   * @return the cleaned-up Prometheus compliant string.
+   */
+  private static String cleanUpString(String string) {
+    return SANITIZE_LEADING_UNDERSCORES
+        .matcher(
+            SANITIZE_TRAILING_UNDERSCORES
+                .matcher(
+                    SANITIZE_CONSECUTIVE_UNDERSCORES
+                        .matcher(INVALID_CHARACTERS_PATTERN.matcher(string).replaceAll("_"))
```
Comment (mateuszrzeszutek) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java:104]: Nit: as a future performance improvement we could probably replace all these regexes (and those in `NameSanitizer` too) with a single loop that removes all the unwanted characters.

**Review by mateuszrzeszutek — APPROVED** [review]

**Review by jsuereth — APPROVED** [review]

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Serializer.java [hunk 9]
```diff
@@ -650,11 +651,28 @@ static Collection<? extends PointData> getPoints(MetricData metricData) {
...
+    // special case - counter
+    if (type == PrometheusType.COUNTER && !name.contains("total")) {
       name = name + "_total";
     }
+    // special case - gauge
+    if (rawMetric.getUnit().equals("1")
+        && type == PrometheusType.GAUGE
+        && !name.contains("ratio")) {
+      name = name + "_ratio";
+    }
```
Comment (jack-berg) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Serializer.java:674]: I don't see this special case ratio rule defined in the spec. What's the source of this?

File: exporters/prometheus/src/test/java/io/opentelemetry/exporter/prometheus/SerializerTest.java [hunk 10]
```diff
@@ -515,4 +596,43 @@ private static String serializeOpenMetrics(MetricData... metrics) {
       throw new UncheckedIOException(e);
     }
   }
+
+  private static MetricData sampleMetricDataGenerator(
```
Comment (jack-berg) [inline exporters/prometheus/src/test/java/io/opentelemetry/exporter/prometheus/SerializerTest.java:600]: #nit: consider renaming to `createSampleMetricData`

File: exporters/prometheus/src/test/java/io/opentelemetry/exporter/prometheus/SerializerTest.java [hunk 11]
```diff
@@ -387,19 +392,19 @@ void prometheus004() {
                 + "monotonic_cumulative_double_sum_suffix_total{otel_scope_name=\"full\",otel_scope_version=\"version\",type=\"mcds\"} 5.0 1633950672000\n"
                 + "# TYPE non_monotonic_cumulative_double_sum gauge\n"
                 + "# HELP non_monotonic_cumulative_double_sum description\n"
-                + "non_monotonic_cumulative_double_sum{otel_scope_name=\"full\",otel_scope_version=\"version\",type=\"nmcds\"} 5.0 1633950672000\n"
+                + "non_monotonic_cumulative_double_sum_ratio{otel_scope_name=\"full\",otel_scope_version=\"version\",type=\"nmcds\"} 5.0 1633950672000\n"
```
Comment (jack-berg) [inline exporters/prometheus/src/test/java/io/opentelemetry/exporter/prometheus/SerializerTest.java:395]: #nit: let's change the unit of `NON_MONOTONIC_CUMULATIVE_LONG_SUM` (which is the instrument that produces this) to something other than `1`. Its strange to an instrument which starts with an up down counter have a `_ratio` suffix, since gauges are most likely to represent things which are ratios.

**Review by jack-berg — COMMENTED** [review]
Couple of small comments but overall looks good. Test cases really help make it clear what's happening. 

Sorry it took so long to get eyes on this!

**Comment by psx95** [comment]
> Couple of small comments but overall looks good. Test cases really help make it clear what's happening.
> 
> Sorry it took so long to get eyes on this!

@jack-berg Thanks for the review !

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusMetricNameMapper.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+   * all of them are used to compute the prometheus equivalent name.
+   *
+   * @param metricData the metric data for which the mapping is to be created.
+   * @param prometheusType the prometheus type to which the metric is to be mapped.
+   * @return a String that acts as the key for mapping between metric data and its prometheus
+   *     equivalent name.
+   */
+  private static String createKeyForCacheMapping(
+      MetricData metricData, PrometheusType prometheusType) {
+    return metricData.getName() + metricData.getUnit() + prometheusType.name();
```
Comment (jack-berg) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusMetricNameMapper.java:73]: Let's create an autovalue class composed of name, unit, and prometheus type to represent the cache key. String concatenation will work most of the time, but there are edge cases where a different combination of name, unit, and prometheus type will yield the same concatenated value. An autovalue class will ensure correct equality.

**Review by jack-berg — APPROVED** [review]
